Authorization decorator

aaronshiel opened the pull request:
Authorization decorator can be added to any endpoint to determine if the person making the request has authorization to manage content (either a content manager or an admin).

Paired with: https://github.com/mentorpal/mentor-graphql/pull/73

mbsimonovic commented on mentor_upload_api/src/mentor_upload_api/authorization_decorator.py:
```diff
@@ -0,0 +1,54 @@
+def authorize_to_manage_content(f):
+    @wraps(f)
+    def authorized_endpoint(*args, **kws):
+        cookies = request.cookies
```
cookies? this is an api, so it should work with jwt/token authentication, right?

mbsimonovic commented on mentor_upload_api/src/mentor_upload_api/authorization_decorator.py:
```diff
@@ -0,0 +1,54 @@
+        if "errors" in tdjson:
+            raise Exception(json.dumps(tdjson.get("errors")))
+        if (
+            "data" not in tdjson
+            or "me" not in tdjson["data"]
+            or "canManageContent" not in tdjson["data"]["me"]
```
json validate response? :)

aaronshiel replied:
will update these in json validation PR

aaronshiel commented:
Key changes:
- merged JSON validation branch with authentication branch
- added JSON validation decorators for uploader entrypoints
- added JSON validation for all graphql queries made by the uploader

mbsimonovic commented on mentor_upload_api/src/mentor_upload_api/authorization_decorator.py:
```diff
@@ -0,0 +1,74 @@
+    @wraps(f)
+    def authorized_endpoint(*args, **kws):
+        bearer_token = request.headers.get("Authorization", "")
+        token_authentication = bool(bearer_token.split(" ")[0].lower() == "bearer")
```
why not simply `token_authentication = bearer_token.lower().startswith('bearer')`

mbsimonovic commented on mentor_upload_api/src/mentor_upload_api/authorization_decorator.py:
```diff
@@ -0,0 +1,74 @@
+    def authorized_endpoint(*args, **kws):
+        bearer_token = request.headers.get("Authorization", "")
+        token_authentication = bool(bearer_token.split(" ")[0].lower() == "bearer")
+        if not token_authentication and not request.cookies.get("refreshToken", ""):
+            log.debug("no authentication token provided")
+            abort(401)
+        headers = {"Authorization": bearer_token} if token_authentication else {}
+        cookies = request.cookies if not token_authentication else {}
```
where is `refreshToken` set? if they change the name of that token then this would break. not sure how likely is that to happen?

mbsimonovic commented on mentor_upload_api/src/mentor_upload_api/blueprints/upload/answer.py:
```diff
@@ -106,19 +125,34 @@ def trim_existing_upload():
+video_upload_json_schema = {
+    "type": "object",
+    "properties": {
+        # Mongoose ObjectID == 24 characters
+        "mentor": {"type": "string", "maxLength": 24, "minLength": 24},
+        "question": {"type": "string", "maxLength": 24, "minLength": 24},
```
mongo is an implementation detail, if graphql changes storage and starts using uuids this will fail, so you're coupling this service to implementation details of another. i'd suggest to just go with strings of length say 5-6 till 40

mbsimonovic commented on mentor_upload_api/src/mentor_upload_api/blueprints/upload/answer.py:
```diff
@@ -64,13 +66,30 @@ def begin_tasks_in_parallel(req):
 @answer_blueprint.route("/trim_existing_upload/", methods=["POST"])
 @answer_blueprint.route("/trim_existing_upload", methods=["POST"])
-def trim_existing_upload():
+@validate_payload_json_decorator(json_schema=trim_existing_upload_json_schema)
+def trim_existing_upload(body):
```
on every trim request the uploader must pay a performance hit and make a graphql api call. If graphql is having issues than this will also fail. But the benefit is that authz uses the last updated permissions. I've usually done this by encoding roles in JWT so no additional http requests are required. The downside is that permissions are cached during the jwt token validation.
